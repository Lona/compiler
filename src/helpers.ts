import { IFS } from 'buffs'
import { createModule, ModuleContext } from './logic/module'
import { Config, load } from './utils/config'
import { defaultReporter, Reporter } from './utils/reporter'

/**
 * Helpers passed to every plugins. They contain:
 *
 * - methods abstracting the file system
 * - a centralized log reporter
 * - the workspace's configuration, etc.
 */
export type Helpers = {
  fs: IFS
  reporter: Reporter
  config: Config
  module: ModuleContext
}

export default async (
  fs: IFS,
  workspacePath: string,
  options: {
    outputPath?: unknown
    reporter?: Reporter
  } = {}
): Promise<Helpers> => {
  const { reporter = defaultReporter } = options

  const helpers: Helpers = {
    fs,
    reporter,
    config: await load(fs, workspacePath),
    module: createModule(fs, workspacePath),
  }

  return helpers
}
