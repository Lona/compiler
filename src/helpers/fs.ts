import * as fs from 'fs'
import * as path from 'path'

export type FSWrapper = {
  readFile(filePath: string): Promise<string>
  writeFile(filePath: string, data: string): Promise<void>
  copyDir(dirPath: string, output?: string): Promise<void>
}

export const createFSWrapper = (
  workspacePath: string,
  _outputPath?: unknown
): FSWrapper => {
  const outputPath =
    typeof _outputPath === 'string'
      ? _outputPath
      : path.join(process.cwd(), 'lona-generated')
  const fsWrapper = {
    readFile(filePath: string) {
      return fs.promises.readFile(
        path.resolve(workspacePath, filePath),
        'utf-8'
      )
    },
    writeFile(filePath: string, data: string) {
      const resolvedPath = path.resolve(outputPath, filePath)
      fs.mkdirSync(path.dirname(resolvedPath), { recursive: true })
      return fs.promises.writeFile(resolvedPath, data, 'utf-8')
    },
    async copyDir(dirPath: string, output: string = '.') {
      const resolvedPath = path.resolve(workspacePath, dirPath)
      const files = await fs.promises.readdir(resolvedPath)

      await Promise.all(
        files.map(async x => {
          if (
            (await fs.promises.stat(path.join(resolvedPath, x))).isDirectory()
          ) {
            return
          }

          return fsWrapper.writeFile(
            path.join(output, x),
            await fsWrapper.readFile(path.join(dirPath, x))
          )
        })
      )
    },
  }

  return fsWrapper
}
