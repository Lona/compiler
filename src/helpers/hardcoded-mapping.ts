// ⚠️ THIS FILE IS AUTO GENERATED. DO NOT MODIFY IT.

import { AST as LogicAST, flattenedMemberExpression } from './logic-ast'
import { EvaluationContext } from './logic-evaluate'

export type HardcodedMap<T, U extends any[]> = {
  functionCallExpression: {
    'Color.setHue': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Color.setSaturation': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Color.setLightness': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Color.fromHSL': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Color.saturate': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Boolean.or': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Boolean.and': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Number.range': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'String.concat': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Array.at': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Optional.value': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'Optional.none': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    Shadow: (node: LogicAST.FunctionCallExpression, ...args: U) => T | undefined
    'FontWeight.ultraLight': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.thin': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.light': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.regular': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.medium': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.semibold': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.bold': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.heavy': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.black': (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
    TextStyle: (
      node: LogicAST.FunctionCallExpression,
      ...args: U
    ) => T | undefined
  }
  memberExpression: {
    'FontWeight.w100': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w200': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w300': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w400': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w500': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w600': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w700': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w800': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
    'FontWeight.w900': (
      node: LogicAST.MemberExpression,
      ...args: U
    ) => T | undefined
  }
}

export const isHardcodedMapCall = {
  functionCallExpression: <T, U extends any[]>(
    k: string,
    map: HardcodedMap<T, U>
  ): k is keyof HardcodedMap<T, U>['functionCallExpression'] => {
    return k in map.functionCallExpression
  },
  memberExpression: <T, U extends any[]>(
    k: string,
    map: HardcodedMap<T, U>
  ): k is keyof HardcodedMap<T, U>['memberExpression'] => {
    return k in map.memberExpression
  },
}

export const createStandardLibraryResolver = <T, U extends any[]>(
  hardcodedMap: HardcodedMap<T, U>
) => (
  node: LogicAST.SyntaxNode,
  evaluationContext: undefined | EvaluationContext,
  ...args: U
): T | undefined => {
  if (!evaluationContext) {
    return undefined
  }

  let matchedHardcodedNode: T | undefined

  if (node.type === 'functionCallExpression') {
    let memberExpression = node.data.expression

    if (!evaluationContext.isFromStandardLibrary(memberExpression.data.id)) {
      return
    }

    const path = (flattenedMemberExpression(memberExpression) || [])
      .map(y => y.string)
      .join('.')

    if (isHardcodedMapCall.functionCallExpression(path, hardcodedMap)) {
      matchedHardcodedNode = hardcodedMap.functionCallExpression[path](
        node,
        ...args
      )
    }
  }

  if (node.type === 'memberExpression') {
    let memberExpression = node

    if (!evaluationContext.isFromStandardLibrary(memberExpression.data.id)) {
      return
    }

    const path = (flattenedMemberExpression(memberExpression) || [])
      .map(y => y.string)
      .join('.')

    if (isHardcodedMapCall.memberExpression(path, hardcodedMap)) {
      matchedHardcodedNode = hardcodedMap.memberExpression[path](node, ...args)
    }
  }

  return matchedHardcodedNode
}
