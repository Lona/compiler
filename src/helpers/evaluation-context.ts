import * as fs from 'fs'
import * as path from 'path'
import { Config } from '../utils/config'
import * as LogicAST from './logic-ast'
import * as LogicScope from './logic-scope'
import * as LogicUnify from './logic-unify'
import * as LogicEvaluate from './logic-evaluate'
import { Reporter } from './reporter'
import { uuid, nonNullable } from '../utils'

/**
 * Iterates through the import declaration of the program
 * and "resolve them", eg. returns a program that contains both
 * the content of imported files and the current program
 */
function resolveImports(
  program: LogicAST.AST.Program,
  reporter: Reporter,
  existingImports: string[] = []
): LogicAST.AST.Program {
  return {
    type: 'program',
    data: {
      id: uuid(),
      block: program.data.block
        .map(x => {
          if (x.type !== 'declaration') {
            return [x]
          }
          if (x.data.content.type !== 'importDeclaration') {
            return [x]
          }

          const libraryName = x.data.content.data.name.name

          if (existingImports.indexOf(libraryName) !== -1) {
            return [x]
          }

          const libraryPath = path.join(
            __dirname,
            '../../static/logic',
            `${libraryName}.logic`
          )
          const libraryExists = fs.existsSync(libraryPath)

          if (!libraryExists) {
            reporter.error(
              `Failed to find library ${libraryName} at path ${libraryPath}`
            )
            return [x]
          }

          const library = LogicAST.makeProgram(
            JSON.parse(fs.readFileSync(libraryPath, 'utf8'))
          )

          if (!library) {
            reporter.error(`Failed to import library ${libraryName}`)
            return [x]
          }

          const resolvedLibrary = resolveImports(
            library,
            reporter,
            existingImports.concat(libraryName)
          )

          return [x, ...resolvedLibrary.data.block]
        })
        .reduce((prev, x) => prev.concat(x), []),
    },
  }
}

export const generate = (config: Config, reporter: Reporter) => {
  const standardLibsPath = path.join(__dirname, '../../static/logic')
  const standardLibs = fs.readdirSync(standardLibsPath)

  const libraryFiles: LogicAST.AST.Program[] = standardLibs.map(
    x =>
      LogicAST.makeProgram(
        JSON.parse(fs.readFileSync(path.join(standardLibsPath, x), 'utf8'))
      ) as LogicAST.AST.Program
  )

  const standardLibsProgram = LogicAST.joinPrograms(libraryFiles)

  const logicPrograms = Object.keys(config.logicFiles)
    .map(k => {
      const node = LogicAST.makeProgram(config.logicFiles[k])
      if (!node) {
        return undefined
      }
      return {
        in: k,
        node,
      }
    })
    .filter(nonNullable)

  const scopeContext = LogicScope.build(
    [{ node: standardLibsProgram, in: 'standard library' }].concat(
      logicPrograms
    ),
    reporter
  )

  const programNode = LogicAST.joinPrograms([
    standardLibsProgram,
    ...logicPrograms.map(x => x.node),
  ])

  const unificationContext = LogicUnify.makeUnificationContext(
    programNode,
    scopeContext,
    reporter
  )
  const substitution = LogicUnify.unify(
    unificationContext.constraints,
    reporter
  )

  const evaluationContext = LogicEvaluate.evaluate(
    programNode,
    programNode,
    scopeContext,
    unificationContext,
    substitution,
    reporter
  )

  return evaluationContext
}
