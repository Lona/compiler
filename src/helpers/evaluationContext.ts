import * as fs from 'fs'
import * as path from 'path'
import { Config } from '../utils/config'
import * as LogicAST from './logicAst'
import * as LogicScope from './logicScope'
import * as LogicUnify from './logicUnify'
import * as LogicEvaluate from './logicEvaluate'
import { Reporter } from '../utils/reporter'
import { nonNullable } from '../utils'

export const STANDARD_LIBRARY = 'standard library'

export const generate = (config: Config, reporter: Reporter) => {
  const standardLibsPath = path.join(__dirname, '../../static/logic')
  const standardLibs = fs.readdirSync(standardLibsPath)

  const libraryFiles: LogicAST.AST.Program[] = standardLibs.map(
    x =>
      LogicAST.makeProgram(
        JSON.parse(fs.readFileSync(path.join(standardLibsPath, x), 'utf8'))
      ) as LogicAST.AST.Program
  )

  const standardLibsProgram = LogicAST.joinPrograms(libraryFiles)

  const logicPrograms = Object.keys(config.logicFiles)
    .map(k => {
      const node = LogicAST.makeProgram(config.logicFiles[k])
      if (!node) {
        return undefined
      }
      return {
        in: k,
        node,
      }
    })
    .filter(nonNullable)

  const scopeContext = LogicScope.build(
    [{ node: standardLibsProgram, in: STANDARD_LIBRARY }].concat(logicPrograms),
    reporter
  )

  const programNode = LogicAST.joinPrograms([
    standardLibsProgram,
    ...logicPrograms.map(x => x.node),
  ])

  const unificationContext = LogicUnify.makeUnificationContext(
    programNode,
    scopeContext,
    reporter
  )
  const substitution = LogicUnify.unify(
    unificationContext.constraints,
    reporter
  )

  const evaluationContext = LogicEvaluate.evaluate(
    programNode,
    programNode,
    scopeContext,
    unificationContext,
    substitution,
    reporter
  )

  return evaluationContext
}
