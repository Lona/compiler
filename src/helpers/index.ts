import { LogicAST } from '@lona/serialization'
import { config as Config } from '../utils'
import { generate as generateEvaluationContext } from './evaluation-context'
import { EvaluationContext } from './logic-evaluate'
import { createStandardLibraryResolver } from './hardcoded-mapping'
import { Reporter, defaultReporter } from './reporter'
import { FSWrapper, createFSWrapper } from './fs'
export { HardcodedMap } from './hardcoded-mapping'
export { EvaluationContext } from './logic-evaluate'
import { TraversalConfig, reduce } from './logic-traversal'
import {
  makeProgram,
  declarationPathTo,
  findParentNode,
  findNode,
} from './logic-ast'

/**
 * Helpers passed to every plugins. They contains some methods abstracting
 * the file system, the evaluation context of the workspace,
 * a reporter (to centralize the logs), the workspace's configuration, etc.
 */
export type Helpers = {
  fs: FSWrapper
  reporter: Reporter
  config: Config.Config
  /**
   * The evaluation context of the Lona Workspace.
   *
   * It contains some mapping between the identifiers and what they reference,
   * eg. `Optional.none` -> the function declaration in Prelude.logic
   * which is used to evaluate `Optional.none` to
   * `{ type: unit, memory: { type: 'unit' } }` for example. It takes care of
   * the scope of the identifier to determine the reference.
   *
   * Additionally, it keeps track of where declaration are from.
   */
  evaluationContext: EvaluationContext | undefined
  /**
   * Takes a hardcoded map `standard library -> node` and creates a function
   * which returns a node if the Logic node passed as argument is from
   * the standard library.
   *
   * Depending on the context, the hardcoded map might use the
   * evaluation context to evaluate the Logic node instead of trying to translate
   * the standard library method into the target format.
   */
  createStandardLibraryResolver: typeof createStandardLibraryResolver
  ast: {
    /** Tries to make a program out of any Logic node */
    makeProgram: (node: LogicAST.SyntaxNode) => LogicAST.Program | undefined
    /** A set of methods to help traverse the AST */
    traversal: {
      declarationPathTo: (
        node: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
        id: string
      ) => string[]
      /** Find the node with the matching id */
      findNode: (
        rootNode: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
        id: string
      ) =>
        | LogicAST.SyntaxNode
        | LogicAST.Pattern
        | LogicAST.Identifier
        | undefined
      /** Find the parent of the node with the matching id */
      findParentNode: (
        rootNode: LogicAST.SyntaxNode | LogicAST.Pattern | LogicAST.Identifier,
        id: string
      ) => LogicAST.SyntaxNode | undefined
      /**
       * The `reduce()` method executes a reducer function (that you provide)
       * on each node of the AST, resulting in a single output value.
       *
       * Your reducer function's returned value is assigned to the accumulator,
       * whose value is remembered across each iteration throughout the AST,
       * and ultimately becomes the final, single resulting value.
       *
       * The traversal is depth-first, you can specify whether it should be
       * pre-order or post-order in the config.
       */
      reduce: <T>(
        node: LogicAST.SyntaxNode,
        callbackfn: (
          previousValue: T,
          currentNode: LogicAST.SyntaxNode,
          config: TraversalConfig
        ) => T,
        initialResult: T,
        config?: TraversalConfig
      ) => T
    }
  }
}

export type PreludeFlags = {
  [id: string]:
    | LogicAST.RecordDeclaration
    | LogicAST.EnumerationDeclaration
    | LogicAST.NamespaceDeclaration
}

export default async (
  workspacePath: string,
  outputPath?: unknown,
  _reporter?: Reporter
): Promise<Helpers> => {
  const fsWrapper = createFSWrapper(workspacePath, outputPath)

  const config = await Config.load(workspacePath, {
    forEvaluation: true,
    fs: fsWrapper,
  })

  let cachedEvaluationContext: EvaluationContext | undefined

  const reporter = _reporter || defaultReporter

  return {
    fs: fsWrapper,
    config,
    get evaluationContext() {
      if (cachedEvaluationContext) {
        return cachedEvaluationContext
      }
      cachedEvaluationContext = generateEvaluationContext(config, reporter)
      return cachedEvaluationContext
    },
    createStandardLibraryResolver,
    reporter,
    ast: {
      makeProgram,
      traversal: {
        declarationPathTo,
        findNode,
        findParentNode,
        reduce,
      },
    },
  }
}
