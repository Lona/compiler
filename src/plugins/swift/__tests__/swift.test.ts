import path from 'path'
import fs from 'fs'
import { createFs, copy, toJSON, match } from 'buffs'
import plugin from '../index'
import { createHelpers } from '../../../helpers'

const workspacePath = path.join(__dirname, '../../../../examples/workspace')
const fixturesPath = path.join(__dirname, '../../__tests__/fixtures')

const tokensBlock = (string: string) => '```tokens\n' + string + '\n```\n'

// TODO: Test more kinds of Swift
describe('Swift', () => {
  it('generates swift', async () => {
    const source = createFs({
      'lona.json': JSON.stringify({}),
      'Colors.md': tokensBlock(`let color: Color = #color(css: "pink")`),
    })

    const helpers = createHelpers(source, '/')

    await plugin.convertWorkspace('/', helpers, { output: '/output' })

    const colors = source.readFileSync('/output/Colors.swift', 'utf8')

    expect(colors).toMatchSnapshot()
  })

  it('converts workspace', async () => {
    const source = createFs()

    copy(fs, source, workspacePath, '/')

    const helpers = createHelpers(source, '/')

    await plugin.convertWorkspace('/', helpers, { output: '/output' })

    const files = toJSON(source, '/output')

    // No need to test helper files
    Object.keys(files).forEach(file => {
      if (file.startsWith('/output/lona-helpers/')) {
        delete files[file]
      }
    })

    expect(files).toMatchSnapshot()
  })

  describe('Fixtures', () => {
    it('converts', async () => {
      const fixtures = match(fs, fixturesPath, { includePatterns: ['**/*.md'] })

      for (let fixture of fixtures) {
        const source = createFs({
          'lona.json': JSON.stringify({}),
          'Fixture.md': fs.readFileSync(
            path.join(fixturesPath, fixture),
            'utf8'
          ),
        })

        const helpers = createHelpers(source, '/')

        await plugin.convertWorkspace('/', helpers, { output: '/output' })

        expect(
          source.readFileSync('/output/Fixture.swift', 'utf8')
        ).toMatchSnapshot(fixture)
      }
    })
  })
})
