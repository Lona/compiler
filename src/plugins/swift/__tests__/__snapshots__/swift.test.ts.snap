// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swift Fixtures converts: declaration/enumeration.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public indirect struct Foo: Wrapped {
  case value(Wrapped)
  case none
}
"
`;

exports[`Swift Fixtures converts: declaration/function.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

func addA(a: T, b: T = 0) -> T {}
"
`;

exports[`Swift Fixtures converts: declaration/import.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif


"
`;

exports[`Swift Fixtures converts: declaration/namespace.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public indirect struct Boolean {
  func foo(a: Bool, b: Bool) -> Bool {}
  func bar(a: Bool, b: Bool) -> Bool {}
}
"
`;

exports[`Swift Fixtures converts: declaration/record.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public struct ThemedColor: Equatable {
  public init(
    light: Color = #colorLiteral(red: 1, green: 1, blue: 1, alpha: 1),
    dark: Color = #colorLiteral(red: 0, green: 0, blue: 0, alpha: 1))
  {
    self.light = light
    self.dark = dark
  }
  public let light: Color = Color(named: \\"light\\")!
  public let dark: Color = Color(named: \\"dark\\")!
}
"
`;

exports[`Swift Fixtures converts: declaration/variable.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public let x: Array = [42]
"
`;

exports[`Swift Fixtures converts: expression/assignment.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

func assignment(a: CGFloat) -> CGFloat {
  a = 10
}
"
`;

exports[`Swift Fixtures converts: expression/functionCall.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

func colorIdentity(a: Color) -> Color {
  return a
}
public let pinkWithDetour: Color = Color(named: \\"pinkWithDetour\\")!
"
`;

exports[`Swift Fixtures converts: expression/identifier.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public let x: CGFloat = 4
func identifier(a: CGFloat) -> CGFloat {
  a = x
}
"
`;

exports[`Swift Fixtures converts: expression/member.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public indirect struct Foo {
  static public let primary: Color = Color(named: \\"primary\\")!
}
public let b: Color = Foo.primary
"
`;

exports[`Swift Fixtures converts: literal/array.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public let x: Array = [42, 35]
"
`;

exports[`Swift Fixtures converts: literal/boolean.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public let trueBoolean: Bool = true
public let falseBoolean: Bool = false
"
`;

exports[`Swift Fixtures converts: literal/color.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public let color: Color = Color(named: \\"color\\")!
public let hexColor: Color = Color(named: \\"hexColor\\")!
"
`;

exports[`Swift Fixtures converts: literal/number.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public let positive: CGFloat = 1
public let float: CGFloat = 0.1
public let negative: CGFloat = -1
"
`;

exports[`Swift Fixtures converts: literal/string.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public let string: String = \\"hello\\"
public let stringWithQuote: String = \\"Hello \\\\\\"world\\\\\\"\\"
"
`;

exports[`Swift Fixtures converts: statement/branch.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

func branchStatement() -> CGFloat {
  if true {
    return 0
  }
  return 1
}
"
`;

exports[`Swift Fixtures converts: statement/loop.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

func loopStatement() -> CGFloat {
  while false {
    return 0
  }
}
"
`;

exports[`Swift Fixtures converts: statement/return.md 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

func returnStatement() -> CGFloat {
  return 0
}
"
`;

exports[`Swift converts workspace 1`] = `
Object {
  "/output/Assets.xcassets/Contents.json": "{
  \\"info\\": {
    \\"version\\": 1,
    \\"author\\": \\"Lona\\"
  }
}",
  "/output/Assets.xcassets/colors/Contents.json": "{
  \\"info\\": {
    \\"version\\": 1,
    \\"author\\": \\"Lona\\"
  }
}",
  "/output/Assets.xcassets/colors/primary.colorset/Contents.json": "{
  \\"info\\": {
    \\"version\\": 1,
    \\"author\\": \\"Lona\\"
  },
  \\"colors\\": [
    {
      \\"idiom\\": \\"universal\\",
      \\"color\\": {
        \\"color-space\\": \\"srgb\\",
        \\"components\\": {
          \\"red\\": \\"0x45\\",
          \\"alpha\\": \\"1.000\\",
          \\"blue\\": \\"0xFF\\",
          \\"green\\": \\"0xCB\\"
        }
      }
    }
  ]
}",
  "/output/Assets.xcassets/colors/testSaturate.colorset/Contents.json": "{
  \\"info\\": {
    \\"version\\": 1,
    \\"author\\": \\"Lona\\"
  },
  \\"colors\\": [
    {
      \\"idiom\\": \\"universal\\",
      \\"color\\": {
        \\"color-space\\": \\"srgb\\",
        \\"components\\": {
          \\"red\\": \\"0x29\\",
          \\"alpha\\": \\"1.000\\",
          \\"blue\\": \\"0xFF\\",
          \\"green\\": \\"0xD7\\"
        }
      }
    }
  ]
}",
  "/output/Colors.swift": "import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public let primary: Color = Color(named: \\"primary\\")!

public let accent: Color = primary

public let testSaturate: Color = Color(named: \\"testSaturate\\")!
",
  "/output/Shadows.swift": "import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public let small: Shadow = Shadow(
  x: 0,
  y: 2,
  blur: 2,
  radius: 0,
  color: #colorLiteral(red: 0.27058823529411763, green: 0.796078431372549, blue: 1, alpha: 1))
",
  "/output/TextStyles.swift": "import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public let heading1: TextStyle = TextStyle(
  fontFamily: \\"Helvetica\\",
  fontSize: 28,
  fontWeight: Font.Weight.bold,
  color: #colorLiteral(red: 0, green: 0.5019607843137255, blue: 0.5019607843137255, alpha: 1))
",
}
`;

exports[`Swift generates swift 1`] = `
"import Foundation

#if canImport(UIKit)
  import UIKit
#elseif canImport(AppKit)
  import AppKit
#endif

public let color: Color = Color(named: \\"color\\")!
"
`;
